File: kotlin/src/main/kotlin/species/CreatorSpeciesResolver.kt
package species

class CreatorSpeciesResolver(
    private val species: Species,
) {
    private val selfAndDescendantsCache = mutableMapOf<String, Set<String>>()
    private val mostSpecies: Specie = species.getByName("Most species") // grep-assumed-does-specie-when-artisan-has-only-doesnt
    private val other: Specie = species.getByName("Other") // grep-species-other

    fun resolveDoes(speciesDoes: Collection<String>, speciesDoesnt: Collection<String>): Set<String> {
        val assumedSpeciesDoes = if (speciesDoes.isEmpty() && speciesDoesnt.isNotEmpty()) {
            setOf(mostSpecies.name)
        } else {
            speciesDoes
        }

        val ordered = getOrderedDoesDoesnt(assumedSpeciesDoes, speciesDoesnt)

        val result = mutableSetOf<String>()

        ordered.forEach { (specie, does) ->
            val descendants = getVisibleSelfAndDescendants(specie)

            if (does) {
                descendants.forEach(result::add)
            } else {
                descendants.forEach(result::remove)
            }
        }

        return result
    }

    /**
     * @return Specie => Does?
     */
    fun getOrderedDoesDoesnt(speciesDoes: Collection<String>, speciesDoesnt: Collection<String>): Map<Specie, Boolean>
    {
        val knownDoes = speciesDoes.map(::getVisibleSpecieOrParentOrOtherForUnusual).flatten().toSet()
        val knownDoesnt = speciesDoesnt.map(::getVisibleSpecieOrEmptySetForUnusual).flatten().toSet()

        var result: List<Pair<Specie, Boolean>> = listOf<Pair<Specie, Boolean>>()
            .plus(knownDoes.map { specie -> specie to true })
            .plus(knownDoesnt.map { specie -> specie to false })

        result = result.sortedWith { item1: Pair<Specie, Boolean>, item2: Pair<Specie, Boolean> ->
            val depthDiff = item1.first.getDepth() - item2.first.getDepth()

            if (0 != depthDiff) { depthDiff } else {
                if (item2.second) 1 else 0 - if (item1.second) 1 else 0
            }
        }

        return result.toMap()
    }

    private fun getVisibleSelfAndDescendants(self: Specie): Set<String>
    {
        return selfAndDescendantsCache.computeIfAbsent(self.name) {
            self.getSelfAndDescendants().map { it.name }.filter { species.getVisibleNames().contains(it) }.toSet()
        }
    }

    private fun getVisibleSpecieOrParentOrOtherForUnusual(specieName: String): Set<Specie>
    {
        if (!species.hasName(specieName)) {
            return setOf(other)
        }

        val result = mutableSetOf<Specie>()
        val unresolved = mutableSetOf(species.getByName(specieName))

        while (unresolved.isNotEmpty()) {
            val specie = unresolved.first()

            if (specie.getHidden()) {
                unresolved.addAll(specie.getParents())
            } else {
                result.add(specie)
            }

            unresolved.remove(specie)
        }

        if (result.size == 0) {
            throw SpecieException("$specieName is hidden and does not have a single visible parent")
        }

        return result
    }

    private fun getVisibleSpecieOrEmptySetForUnusual(specieName: String): Set<Specie>
    {
        if (!species.hasName(specieName) || species.getByName(specieName).getHidden()) {
            return setOf()
        }

        return setOf(species.getByName(specieName))
    }
}

File: symfony/src/Species/CreatorSpeciesResolver.php
<?php

declare(strict_types=1);

namespace App\Species;

use App\Species\Hierarchy\Specie;
use App\Species\Hierarchy\Species;
use App\Species\Hierarchy\SpecieSet;
use Veelkoov\Debris\Base\DMap;
use Veelkoov\Debris\Lists\StringList;
use Veelkoov\Debris\Maps\Pair;
use Veelkoov\Debris\Sets\StringSet;

final class CreatorSpeciesResolver
{
    /**
     * @var array<string, StringSet>
     */
    private array $selfAndDescendantsCache = [];

    private readonly Specie $mostSpecies;
    private readonly Specie $other;

    public function __construct(
        private readonly Species $species,
    ) {
        $this->mostSpecies = $this->species->getByName('Most species'); // grep-assumed-does-specie-when-creator-has-only-doesnt
        $this->other = $this->species->getByName('Other'); // grep-species-other
    }

    /**
     * Based on the list of names of all the species done by the creator, return list of names
     * of all the species available in the filters, which should match the creator.
     */
    public function resolveForFilters(StringSet $speciesNames): StringSet
    {
        return $this->species->getFlat()->filter(function (Specie $specie) use ($speciesNames) {
            if ($specie->getHidden()) {
                return false;
            }

            $thisAndDescendantsNames = $specie->getThisAndDescendants()->getNames();

            return $speciesNames->any(static fn (string $specieName) => $thisAndDescendantsNames->contains($specieName));
        })->getNames();
    }

    /**
     * Based on the list of species done by the creator and the exceptions ("doesn't"),
     * calculate a list of names of all the species actually done.
     */
    public function resolveDoes(StringList $speciesDoes, StringList $speciesDoesnt): StringSet
    {
        $assumedSpeciesDoes = $speciesDoes->isEmpty() && $speciesDoesnt->isNotEmpty()
            ? StringList::of($this->mostSpecies->getName()) : $speciesDoes;

        $ordered = $this->getOrderedDoesDoesnt($assumedSpeciesDoes, $speciesDoesnt);

        $result = new StringSet();

        foreach ($ordered as $specie => $does) {
            $descendants = $this->getVisibleSelfAndDescendants($specie);

            foreach ($descendants as $descendant) {
                if ($does) {
                    $result->add($descendant);
                } else {
                    $result->remove($descendant);
                }
            }
        }

        return $result;
    }

    /**
     * @return DMap<Specie, bool> Specie => Does?
     */
    public function getOrderedDoesDoesnt(StringList $speciesDoes, StringList $speciesDoesnt): DMap
    {
        /** @var DMap<Specie, bool> $result */
        $result = new DMap();

        foreach ($speciesDoes as $specieDone) {
            foreach ($this->getVisibleSpecieOrParentOrOtherForUnusual($specieDone) as $specie) {
                $result->set($specie, true);
            }
        }

        foreach ($speciesDoesnt as $specieNotDone) {
            foreach ($this->getVisibleSpecieOrEmptySetForUnusual($specieNotDone) as $specie) {
                $result->set($specie, false);
            }
        }

        return $result->sorted(function (Pair $item1, Pair $item2): int {
            $depthDiff = $item1->key->getDepth() - $item2->key->getDepth();

            if (0 !== $depthDiff) {
                return $depthDiff;
            } elseif ($item2->value) {
                return 1;
            } else {
                return 0 - ($item1->value ? 1 : 0);
            }
        });
    }

    private function getVisibleSelfAndDescendants(Specie $specie): StringSet
    {
        return $this->selfAndDescendantsCache[$specie->getName()] ??= $specie->getThisAndDescendants()
            ->filter(static fn (Specie $specie) => !$specie->getHidden())->getNames();
    }

    private function getVisibleSpecieOrParentOrOtherForUnusual(string $specieName): SpecieSet
    {
        if (!$this->species->hasName($specieName)) {
            return SpecieSet::of($this->other);
        }

        $result = new SpecieSet();
        $unresolved = [$this->species->getByName($specieName)];

        while ([] !== $unresolved) {
            $specie = array_shift($unresolved);

            if ($specie->getHidden()) {
                array_push($unresolved, ...$specie->getParents());
            } else {
                $result->add($specie);
            }
        }

        if ($result->isEmpty()) {
            throw new SpecieException("$specieName is hidden and does not have a single visible parent");
        }

        return $result;
    }

    private function getVisibleSpecieOrEmptySetForUnusual(string $specieName): SpecieSet
    {
        if (!$this->species->hasName($specieName) || $this->species->getByName($specieName)->getHidden()) {
            return new SpecieSet();
        }

        return SpecieSet::of($this->species->getByName($specieName));
    }
}
